pkg/trillianclient/trillian_client.go

```diff
@@ -229,7 +229,42 @@ func (t *TrillianClient) GetLeafAndProofByHash(ctx context.Context, hash []byte)
 		}
 	}
 
-	return t.GetLeafAndProofByIndex(ctx, proofs[0].LeafIndex)
+	leafIndex := proofs[0].LeafIndex
+	// fetch the leaf without re-requesting a proof (since we already have it)
+	leafOnlyResp := t.GetLeafWithoutProof(ctx, leafIndex)
+	if leafOnlyResp.Err != nil {
+		return &Response{
+			Status: status.Code(leafOnlyResp.Err),
+			Err:    leafOnlyResp.Err,
+		}
+	}
+
+	if leafOnlyResp.GetLeavesByRangeResult == nil || len(leafOnlyResp.GetLeavesByRangeResult.Leaves) == 0 {
```

Comment:
Can we pull 242 to 258 into its own method to reuse it in AddLeaf?

Reply:
ah yes, good call.